modules/distribution-zones/src/main/java/org/apache/ignite/internal/distributionzones/disaster/DisasterRecoveryManager.java

```diff
@@ -0,0 +1,170 @@
+
+import java.util.UUID;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.TimeUnit;
+import org.apache.ignite.internal.affinity.Assignments;
+import org.apache.ignite.internal.catalog.CatalogManager;
+import org.apache.ignite.internal.distributionzones.DistributionZoneManager;
+import org.apache.ignite.internal.lang.ByteArray;
+import org.apache.ignite.internal.logger.IgniteLogger;
+import org.apache.ignite.internal.logger.Loggers;
+import org.apache.ignite.internal.manager.IgniteComponent;
+import org.apache.ignite.internal.metastorage.Entry;
+import org.apache.ignite.internal.metastorage.MetaStorageManager;
+import org.apache.ignite.internal.metastorage.WatchEvent;
+import org.apache.ignite.internal.metastorage.WatchListener;
+import org.apache.ignite.internal.util.ByteUtils;
+
+/**
+ * Manager, responsible for "disaster recovery" operations.
+ * Internally it triggers meta-storage updates, in order to acquire unique causality token.
+ * As a reaction to these updates, manager performs actual recovery operations, such as {@link #manualGroupsUpdate(int, int)}.
+ */
+public class DisasterRecoveryManager implements IgniteComponent {
+    /** Logger. */
+    private static final IgniteLogger LOG = Loggers.forClass(DisasterRecoveryManager.class);
+
+    /** Single key for writing disaster recovery requests into meta-storage. */
+    static final ByteArray RECOVERY_TRIGGER_KEY = new ByteArray("disaster.recovery.trigger");
+
+    /** Meta-storage manager. */
+    final MetaStorageManager metaStorageManager;
+
+    /** Catalog manager. */
+    final CatalogManager catalogManager;
+
+    /** Distribution zone manager. */
+    final DistributionZoneManager dzManager;
+
+    /** Watch listener for {@link #RECOVERY_TRIGGER_KEY}. */
+    private final WatchListener watchListener;
+
+    /** Map of operations, triggered by local node, that have not yet been processed by {@link #watchListener}. */
+    private final ConcurrentMap<UUID, CompletableFuture<Void>> ongoingOperations = new ConcurrentHashMap<>();
+
+    /**
+     * Constructor.
+     *
+     * @param metaStorageManager Meta-storage manager.
+     * @param catalogManager Catalog manager.
+     * @param dzManager Distribution zone manager.
+     */
+    public DisasterRecoveryManager(
+            MetaStorageManager metaStorageManager,
+            CatalogManager catalogManager,
+            DistributionZoneManager dzManager
+    ) {
+        this.metaStorageManager = metaStorageManager;
+        this.catalogManager = catalogManager;
+        this.dzManager = dzManager;
+
+        watchListener = new WatchListener() {
+            @Override
+            public CompletableFuture<Void> onUpdate(WatchEvent event) {
+                return handleTriggerKeyUpdate(event);
+            }
+
+            @Override
+            public void onError(Throwable e) {
+                // No-op.
```

Comment:
Maybe log error ?

Reply:
I think it should already be logged elsewhere